provider/linode/linode_discover.go

```diff
@@ -0,0 +1,163 @@
+`
+}
+
+func (p *Provider) Addrs(args map[string]string, l *log.Logger) ([]string, error) {
+	if args["provider"] != "linode" {
+		return nil, fmt.Errorf("discover-linode: invalid provider " + args["provider"])
+	}
+
+	if l == nil {
+		l = log.New(ioutil.Discard, "", 0)
+	}
+
+	addressType := args["address_type"]
+	region := args["region"]
+	tagName := args["tag_name"]
+	apiToken := argsOrEnv(args, "api_token", "LINODE_TOKEN")
+	l.Printf("[DEBUG] discover-linode: Using address_type=%s region=%s tag_name=%s", addressType, region, tagName)
+
+	client := getLinodeClient(p.userAgent, apiToken)
+
+	filters := Filter{
+		Region: "",
+		Tag:    "",
+	}
+
+	if region != "" {
+		filters.Region = region
+	}
+	if tagName != "" {
+		filters.Tag = tagName
+	}
+
+	jsonFilters, _ := json.Marshal(filters)
+	filterOpt := linodego.ListOptions{Filter: string(jsonFilters)}
+
+	linodes, err := client.ListInstances(context.Background(), &filterOpt)
+	if err != nil {
+		return nil, fmt.Errorf("discover-linode: Fetching Linode instances failed: %s", err)
+	}
+
+	var addrs []string
+	for _, linode := range linodes {
+		if linodeAddrs, err := getLinodeAddresses(client, addressType, linode); err == nil {
+			addrs = append(addrs, linodeAddrs...)
+		} else {
+			return nil, err
+		}
+	}
+
+	return addrs, nil
+}
+
+func getLinodeAddresses(client linodego.Client, addressType string, linode linodego.Instance) (addresses []string, err error) {
+	switch addressType {
+	case "public_v4":
+		for _, ip := range linode.IPv4 {
+			if !privateIP(*ip) {
+				addresses = append(addresses, ip.String())
+			}
+		}
+	case "public_v6":
+		v6Addr := strings.SplitN(linode.IPv6, "/", 2)
+		if len(v6Addr) > 0 {
+			addresses = append(addresses, v6Addr[0])
+		}
+	case "private_v6":
+		var addr *linodego.InstanceIPAddressResponse
+		if addr, err = client.GetInstanceIPAddresses(context.Background(), linode.ID); err == nil {
+			if addr.IPv6.LinkLocal.Address != "" {
+				addresses = append(addresses, addr.IPv6.LinkLocal.Address)
+			}
+		} else {
+			err = fmt.Errorf("discover-linode: Fetching Linode IP address for instance %v failed: %s", linode.ID, err)
+		}
+	// private_v4
+	default:
+		for _, ip := range linode.IPv4 {
+			if privateIP(*ip) {
+				addresses = append(addresses, ip.String())
+			}
+		}
+	}
+	return addresses, err
+}
+
+func getLinodeClient(userAgent, apiToken string) linodego.Client {
+	tokenSource := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: apiToken})
+
+	oauth2Client := &http.Client{
+		Transport: &oauth2.Transport{
+			Source: tokenSource,
+		},
+	}
+
+	client := linodego.NewClient(oauth2Client)
+	if userAgent != "" {
+		client.SetUserAgent(userAgent)
+	}
+
+	return client
+}
+
+func argsOrEnv(args map[string]string, key, env string) string {
+	if value := args[key]; value != "" {
+		return value
+	}
+	return os.Getenv(env)
+}
+
+// privateIP determines if an IP is for private use (RFC1918)
+// https://stackoverflow.com/a/41273687
+func privateIP(ip net.IP) bool {
```

Comment:
I don't really like this approach. I feel like we're doing it to fit go-discover semantics, instead of just surfacing the Linode API for this.

I'd basically like to see a simpler interface for selecting types that doesn't require this private IP check. It is perfectly fine if we do what is more "obvious" or "expected" to Linode API users than to do what is expected for Consul users.

Does that make sense? Could we try to think about the options for address type more in a way a Linode user would expect and remove the need for this?

Reply:
The [Linode API](https://developers.linode.com/api/docs/v4#operation/getLinodeInstances) returns Instance lists with an array of IPv4 addresses.  In this list, both private and public IP addresses are commingled.  These addresses are all valid for the `eth0` device of the instance.

Linode private IP addresses are in the range of `192.168.0.0/16`.  This function could be reduced to that set, if that's more inline with what you are thinking.  For that matter, a "192.168." string prefix check would suffice.

There is networking endpoint that can be called, which precisely identifies the type of each address, but this endpoint must be called for each instance, which adds time and increases the chances of hitting a rate limit.  https://developers.linode.com/api/docs/v4#operation/getLinodeInstances

Reply:
What @displague said is spot on. I originally did have something more obvious, using the actual networking endpoint for each Linode, as seen here:

https://github.com/hashicorp/go-discover/blob/55ba9e8657a667ca67c09df287ff3038ef5eb9a0/provider/linode/linode_discover.go#L80-L114

It is a balancing act of speed for clarity and conciseness in the code.  I felt this was a happy medium, but if you would rather it be more explicit like the above, let me know. If we go with the explicit networking API calls, the concerns of additional time and possible rate limiting would apply.

Reply:
I'm not that familiar with the rate limiting requirements, but it feels like the API is a more stable approach in the longer term and makes go-discover do less, which I think is preferable. 

If we know the upstream API will rate-limit in standard use-cases I'd feel it would be appropriate to try to work around that (either by respecting some upstream "rate limit" requirements like headers and sleeping) and if necessary this current approach, it just doesn't feel optimal.

I definitely wouldn't be concerned about latency assuming the API isn't _terribly_ slow (eg. many requests measured in seconds). There's an assumed bootstrapping time for Consul and an awareness that it will "eventually" join without any sort-of SLA.

Reply:
So to be clear, I'd prefer the approach you originally did that was inlined there if you agree on that above.

Reply:
Sounds good. Thanks for taking the time to lay out your thoughts. I reverted back to using the features of the API to get these values, eliminating the need for the function to check whether it was private or public.